File: 1442-count-triplets-that-can-form-two-arrays-of-equal-xor/1442-count-triplets-that-can-form-two-arrays-of-equal-xor.java
class Solution {
    public int countTriplets(int[] arr) {
        int n = arr.length, ans = 0;
        int[] prefixXor = new int[n+1];
        prefixXor[0] = 0;
        for(int i=1; i<=n; i++) {
            prefixXor[i] = arr[i-1] ^ prefixXor[i-1];
        }
        for(int i=0; i<=n; i++) {
            for(int j=i+1; j<=n; j++) {
                if((prefixXor[i] ^ prefixXor[j]) == 0) ans += j-i-1;
            }
        }
        return ans;
    }
}

File: 1442-count-triplets-that-can-form-two-arrays-of-equal-xor/1442-count-triplets-that-can-form-two-arrays-of-equal-xor.cpp
class Solution {
public:
    int countTriplets(vector<int>& arr) {
        int n = arr.size();
        unordered_map<int, vector<int>>mp;
        mp[0].push_back(-1);
        int x = 0, ans = 0;
        for(int i=0; i<n; i++) {
            x ^= arr[i];
            if(mp.find(x) != mp.end()) {
                for(auto index:mp[x]) {
                    ans += i - index - 1;
                }
            }
            mp[x].push_back(i);
        }
        return ans;
    }
};